File: matlab/ode.m
clc
clear
close all

lambda = 1.0;
initial_condition = 1.0;
source_term = @(t) -t;

duration = 1.0;
time_step = 0.01;

euler_config = Config(duration, time_step, Method.Euler);
rk4_config = Config(duration, time_step, Method.RK4);

eq = ODE(lambda, initial_condition, source_term);

rk4 = @() eq.solve(rk4_config);
euler = @() eq.solve(euler_config);

bench("ode-rk4", rk4, 1000);
bench("ode-euler", euler, 1000);

File: cpp/src/bin/ode.cpp
import de.ode;
import de.utils;

int main() {
    Config rk4_config{};
    Config euler_config{.method = Method::Euler};
    auto f = [](double t) -> double { return -t; };
    ODE ode{.lambda = 1.0, .initial_condition = 1.0, .source_term = f};

    auto euler_solver = [ode, euler_config]() {
        return ode.solve(euler_config);
    };
    auto rk4_solver = [ode, rk4_config]() { return ode.solve(rk4_config); };

    bench("ode-euler", euler_solver, 10000);
    bench("ode-rk4", rk4_solver, 10000);

    return 0;
}
